File: packages/vscode-shared/src/types/api/languageModels.ts
import {
	LmProviderInfo as DatabaseLmProviderInfo,
	LmProviderBaseModelInfo as DatabaseLmProviderBaseModelInfo,
	LmProviderProperty as DatabaseLmProviderProperty,
} from "../database/LmProviderInfo";

export const LlmFeatures = ["all", "conversational", "vision", "embedding", "tools"] as const;
export type LlmFeature = typeof LlmFeatures[number];

export type LmProviderBaseModelInfo = Omit<DatabaseLmProviderBaseModelInfo, "version" | "entityName">;
export type ListLanguageModelsResponse = {
	basemodels: LmProviderBaseModelInfo[];
};

export type LmProviderProperty = Omit<DatabaseLmProviderProperty, "version" | "entityName">;

export type LmProviderInfoResponse = Omit<DatabaseLmProviderInfo, "version" | "entityName"> & {
	status: string,
}

export type ListLmProvidersResponse = {
	providers: LmProviderInfoResponse[];
};

export type UpdateLmProviderRequest = {
	lmProviderId: string,
	properties: Record<string, string> | null,
	weight: number | null,
    selectedModels: string[] | null,
	embeddingModelIndexes: number[] | null,
	visionModelIndexes: number[] | null,
	toolsModelIndexes: number[] | null,
};

export type UpdateLmProviderResponse = LmProviderInfoResponse;

File: server/src/aif_types/languagemodels.py
from typing import List, Optional
from pydantic import BaseModel
from aif_types.llm import LlmProvider


class LanguageModelInfo(BaseModel):
    provider: LlmProvider
    basemodel_uri: str
    name: str
    ready: bool # for Ollama, if it's not ready, need to download it first
    weight: int


class ListLanguageModelsResponse(BaseModel):
    basemodels: List[LanguageModelInfo]


class LmProviderBaseModelInfo(BaseModel):
    # For the models with no version, this is the model name or deployment name, otherwise it's <model-name>:<version>
    id: str
    selected: bool
	# ONLY for Azure OpenAI currently, we allow the users to add deployment name + version, and they can delete it
    isUserDefined: bool = False
    tags: List[str]


class LmProviderProperty(BaseModel):
    description: str
    hint: str
    value: str | None
    isCredential: bool


class LmProviderEntity(BaseModel):
    lmProviderId: str
    name: str
    properties: dict[str, LmProviderProperty]
    weight: int
    supportUserDefinedModels: bool = False
    models: List[LmProviderBaseModelInfo]
    status: str     # available or unavailable


class ListLmProvidersResponse(BaseModel):
    providers: List[LmProviderEntity]


class UpdateLmProviderRequest(BaseModel):
    lmProviderId: str
    properties: Optional[dict[str, str]] = None
    weight: Optional[int] = None
    selectedModels: Optional[List[str]] = None
    embeddingModelIndexes: Optional[List[int]] = None       # for LmProviderBaseModelInfo with isUserDefined, e.g. Azure OpenAI
    visionModelIndexes: Optional[List[int]] = None          # for LmProviderBaseModelInfo with isUserDefined, e.g. Azure OpenAI
    toolsModelIndexes: Optional[List[int]] = None           # for LmProviderBaseModelInfo with isUserDefined, e.g. Azure OpenAI
